Trend modularization

karl-richter opened the pull request:
## Background

- To reduce complexity in the TimeNet, each component should become its own module on the long run.

## Key changes

- This PR contains the minimal changes to migrate the Trend to a separate component.

## Review Checklist
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, added docstrings and data types to function definitions.
- [x] I have added pytests to check whether my feature / fix works.

Please make sure to follow our best practices in the [Contributing guidelines](https://github.com/ourownstory/neural_prophet/blob/main/CONTRIBUTING.md).

alfonsogarciadecorral commented:
Hey @karl-richter 

Nice changes. Two things from me:
- Should we have a trend class and then derived classes which represent different models/configuration? In that case should the trend base be an abstract class instead?
- I started to think about which is the most modular way to give the combinations of trend_modl_type  and glocal/local/global. I'm not fully sure, I need a little bit more time. Im going to use the Tuesday 27th to go through it and try to have a proposal then :) 

apart from that it all makes sense

karl-richter commented:
> Hey @karl-richter
> 
> Nice changes. Two things from me:
> 
> * Should we have a trend class and then derived classes which represent different models/configuration? In that case should the trend base be an abstract class instead?
> * I started to think about which is the most modular way to give the combinations of trend_modl_type  and glocal/local/global. I'm not fully sure, I need a little bit more time. Im going to use the Tuesday 27th to go through it and try to have a proposal then :)
> 
> apart from that it all makes sense

@alfonsogarciadecorral @ourownstory  
As we discussed, I performed the following changes:
- Added a router function to initialize the correct trend based on multiple conditions (`components/router.py`)
- Separated the pure linear trend from the piecewise linear trend
- Added a static trend (only the bias, no growth)
- Added a test for the trend without changepoints

alfonsogarciadecorral reviewed (approved):
Great stuff!
I just made a small change as we don't need the `get_trend_deltas` property for the StaticTrend

ourownstory commented on neuralprophet/components/base.py:
```diff
@@ -0,0 +1,35 @@
+        self.id_list = id_list
+        self.device = device
+
+    def forward(self, x):
```
`@abstract`

ourownstory replied:
alternative to having separate Global/Local/Glocal classes:
introduce additional arg
```
if self.glocal == "global":
    return forward_global()
elif ...
```

ourownstory commented on neuralprophet/components/router.py:
```diff
@@ -0,0 +1,69 @@
+    """
+    if config.growth == "off":
+        # No trend
+        return StaticTrend(
```
Please complete docstrings.

We could also parse the function args into a dict and then parse(`**`) that into the appropriate trend classes (as all the args seem to be identical and identically named)

ourownstory commented on neuralprophet/components/base.py:
```diff
@@ -0,0 +1,35 @@
+        """
+        pass
+
+    def new_param(self, dims):
```
Let's make this a util function?

ourownstory commented on neuralprophet/components/trend/trend.py:
```diff
@@ -0,0 +1,53 @@
+        )
+
+    @abstractmethod
+    def forward(self, t, meta):
```
was this already defined in base component?

ourownstory commented on neuralprophet/components/structure.md:
````diff
@@ -0,0 +1,11 @@
+```mermaid
````
convert to wiki

ourownstory commented on neuralprophet/components/trend/linear.py:
```diff
@@ -0,0 +1,105 @@
+from neuralprophet.components.trend import Trend
+
+
+class LinearTrend(Trend):
```
evaluate using dataclasses (for all comonents)

ourownstory reviewed (approved):
Great Work!
Jut added minor comments, feel free to merge once you are happy with how you addressed them.
